Reduce load time of trace page by deferring critical path tooltip

DamianMaslanka5 opened the pull request:
~I am opening this PR without an issue, I understand that is might not be an acceptable change, feel free to close this PR if you don't agree with this change.~

Improve performance (~300ms for 500 spans) by loading tooltip only on hover

* Related issue in antd repo https://github.com/ant-design/ant-design/issues/48468
* My other changes which improve performance related to critical path https://github.com/jaegertracing/jaeger-ui/pull/2716
* Towards https://github.com/jaegertracing/jaeger-ui/issues/645, to improve general performance with large amount of spans

## Before

![criticalPathTooltip](https://github.com/user-attachments/assets/908b6279-a3b8-4446-a08b-386220e22141)


## After

![criticalPathNoTooltip](https://github.com/user-attachments/assets/e7eb14e2-6acd-4f6d-94cb-ce1d82515e02)

yurishkuro reviewed (commented):
I understand what you're trying to do (effectively reduce the size of the DOM), but the resulting behavior is unintuitive, from a user perspective the tooltip would work or not "randomly" (unless they know about 500 spans threshold).

Isn't there another mechanism in React where the tooltip doesn't need to be physically added to the DOM but still can be shown dynamically?

DamianMaslanka5 commented on packages/jaeger-ui/src/components/TracePage/TraceTimelineViewer/SpanBar.tsx:
```diff
@@ -54,6 +54,55 @@ function toPercentInDecimal(value: number) {
+      style={{
+        background: 'black',
+        left: toPercentInDecimal(props.criticalPathViewStart),
+        width: toPercentInDecimal(props.criticalPathViewEnd - props.criticalPathViewStart),
```
**note:** I am ok with a little bit of duplicated code here and in L81, because it is contained in a single component.

But let me know if this a problem, I can add another component (`SpanBarCriticalPathWithoutTooltip`?) to remove the duplication

DamianMaslanka5 commented:
`I understand what you're trying to do (effectively reduce the size of the DOM), but the resulting behavior is unintuitive, from a user perspective the tooltip would work or not "randomly" (unless they know about 500 spans threshold).`

I agree this is not perfect.

`Isn't there another mechanism in React where the tooltip doesn't need to be physically added to the DOM but still can be shown dynamically?`

Thanks for the question, it made me think a bit more. Tbh I am not very familiar with react, so I am open to all suggestions.

---

I made changes to create the tooltip only on hover 41ef333798b14aceaa29f10cf4d746f34bfe297f, I just added another commit, and removed previous changes in the same commit, just remember to squash the commits when merging (or I can squash this locally and force push), I am not sure if you squash all commits by default, on merge

## Render time in profiler looks good

![image](https://github.com/user-attachments/assets/52dac4a6-daea-43d1-a165-b387505a04fd)

## Tooltips work correctly


![firefox_f27lZW4NM5](https://github.com/user-attachments/assets/4f7a553c-08e1-47f5-a072-783cafe5098a)

DamianMaslanka5 commented:
@yurishkuro, I updated this branch

1. I added test to avoid reducing code coverage
2. I removed the initial commit, which disabled the tooltip when 500 spans existed. Now tooltip is loaded only when hovering mouse over critical path.

yurishkuro commented on packages/jaeger-ui/src/components/TracePage/TraceTimelineViewer/SpanBar.tsx:
```diff
@@ -54,6 +54,52 @@ function toPercentInDecimal(value: number) {
+          </div>
+        }
+      >
+        <div
```
can we avoid duplicating this code segment?

DamianMaslanka5 replied:
Yes, see https://github.com/jaegertracing/jaeger-ui/pull/2718#discussion_r2041031917

I will change this tomorrow
